src/node_native_module.cc

```diff
@@ -0,0 +1,323 @@
+  // because here they are only looked up from the C++ side
+  // (except in process.binding('natives') which we don't use)
+  // so there is little value to put them in a v8::Object upfront.
+  // Moreover, a std::map lookup should be faster than a lookup on
+  // an V8 Object in dictionary mode.
+  Isolate* isolate = env->isolate();
+  Local<Context> context = env->context();
+  Local<Object> native_module_source = Object::New(isolate);
+  DefineJavaScript(env, native_module_source);
+  native_module_source->SetIntegrityLevel(context, IntegrityLevel::kFrozen)
+      .FromJust();
+  env->set_native_module_source(native_module_source);
+
+  Local<Object> native_module_source_hash = Object::New(isolate);
+  DefineJavaScriptHash(env, native_module_source_hash);
+  native_module_source_hash->SetIntegrityLevel(context, IntegrityLevel::kFrozen)
+      .FromJust();
+  env->set_native_module_source_hash(native_module_source_hash);
+
+  Local<Object> native_module_code_cache = Object::New(isolate);
+  DefineCodeCache(env, native_module_code_cache);
+  native_module_code_cache->SetIntegrityLevel(context, IntegrityLevel::kFrozen)
+      .FromJust();
+  env->set_native_module_code_cache(native_module_code_cache);
+
+  Local<Object> native_module_code_cache_hash = Object::New(isolate);
+  DefineCodeCacheHash(env, native_module_code_cache_hash);
+  native_module_code_cache_hash
+      ->SetIntegrityLevel(context, IntegrityLevel::kFrozen)
+      .FromJust();
+  env->set_native_module_code_cache_hash(native_module_code_cache_hash);
+
+  env->set_native_module_column_offset(Integer::New(isolate, 0));
+  env->set_native_module_line_offset(Integer::New(isolate, 0));
+
+  env->set_native_module_with_cache(Set::New(isolate));
+  env->set_native_module_without_cache(Set::New(isolate));
+}
+
+void NativeModule::CompileCodeCache(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  CHECK(args[0]->IsString());
+  Local<String> id = args[0].As<String>();
+
+  Local<Value> result = CompileAsModule(env, id, true);
+  if (!result.IsEmpty()) {
+    args.GetReturnValue().Set(result);
+  }
+}
+
+void NativeModule::CompileFunction(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK(args[0]->IsString());
+  Local<String> id = args[0].As<String>();
+
+  Local<Value> result = CompileAsModule(env, id, false);
+  if (!result.IsEmpty()) {
+    args.GetReturnValue().Set(result);
+  }
+}
+
+Local<Value> NativeModule::CompileAsModule(Environment* env,
+                                           Local<String> id,
+                                           bool produce_code_cache) {
+  Local<String> parameters[] = {env->exports_string(),
+                                env->require_string(),
+                                env->module_string(),
+                                env->process_string(),
+                                env->internal_binding_string()};
+
+  return Compile(
```

Comment:
This is private, so we know there are 5 parameters, IMO this seems like a place we should use:
`std::array<Local<String>, 5>`

Reply:
@refack I plan to reuse this for the bootstrappers which have different numbers of parameters - we could template the N here but that would force us to put this chunk of code in the header?

Reply:
Although I could make them `v8::Local<v8::String> parameters[]`, since that's what `ScriptCompiler::CompileFunctionInContext` takes - it's slightly better than a raw pointer

Reply:
ack